packages/datastore/src/datastore.ts

```diff
@@ -92,27 +131,91 @@ class Datastore implements IIterable<Table> {
    * `O(log32 n)`
    */
   get<S extends Schema>(schema: S): Table<S> {
-    throw '';
+    return this._tables[schema.id] as Table<S>;
   }
 
   /**
+   * Begin a new transaction in the store.
+   *
+   * @returns The id of the new transaction
    *
+   * #### Notes
+   * This will allow the state of the store to be mutated
+   * thorugh the `update` method on the individual tables.
+   *
+   * After the updates are completed, `endTransaction` should
+   * be called.
    */
   beginTransaction(): string {
-    throw '';
+    const id = createDuplexId(this.version, this.id);
+    this._initTransaction(id);
+    return id;
   }
 
   /**
+   * Completes a transaction.
    *
+   * #### Notes
+   * This completes a transaction previously started with
+   * `beginTransaction`. If a change has occurred, the
+   * `changed` signal will be emitted.
    */
   endTransaction(): void {
-    throw '';
+    this._finalizeTransaction();
+    if (this._context.change) {
+      this._changed.emit({
+        storeId: this.id,
+        transactionId: this._currentTransactionId,
+        type: 'transaction',
+        change: this._context.change,
+      });
+    }
+    if (this.transactionBroadcastHandler && this._context.patch) {
+      MessageLoop.sendMessage(
+        this.transactionBroadcastHandler,
+        new Datastore.TransactionMessage({
+          id: this._currentTransactionId,
+          storeId: this.id,
+          patch: this._context.patch,
+      }));
+    }
+  }
+
+
+  /**
+   * Apply a transaction to the datastore.
+   *
+   * @param transaction - The data of the transaction.
+   *
+   * @returns A promise which resolves when the action is complete.
+   *
+   * @throws An exception if `apply` is called during a mutation.
+   *
+   * #### Notes
+   * If changes are made, the `changed` signal will be emitted.
+   */
+  apply(transaction: Datastore.Transaction): void {
+    const {storeId, patch} = transaction;
+
+    this._initTransaction(transaction.id);
```

Comment:
This method may throw if you're already in a transaction. You should detect that and hold on to the transaction to process later, otherwise, you'll drop the transaction with no way to recover.

Reply:
So we are assuming that the user might do asynchronous steps between `beginTransaction` and `endTransaction`? That does complicate the logic some. I'll have to think through this in more detail to look for corner cases. I'll ping again once I have a proposed solution.

Reply:
Ok, with the way I did it now: Async steps are not supported (transactions will be automatically closed via message loop), but any *synchronous* code that produce external transactions between start/endTransaction will be queued for processing later. I would say doing so is a bit of an antipattern, but now the datastore will robustly handle it.

The behavior with async steps is undefined, but it will at least produce legible exceptions/warnings when things go wrong.